File: matlab/day15_part2_2018.m

function main
    elf_power = 3;
    final_outcome = 0;
    while true
        elf_power = elf_power + 1;
        [map, units, rows, cols, num_units] = parse_map(elf_power);
        rounds = 0;
        elf_died = false;
        while true
            [combat_continues, hp_sum] = get_status(units, num_units);
            if ~combat_continues
                final_outcome = rounds * hp_sum;
                break
            end
            [round_completed, elf_died, map, units, num_units] = tick(map, units, rows, cols, num_units, true);
            if elf_died
                break
            end
            if ~round_completed && ~elf_died
                [~ , hp_sum] = get_status(units, num_units);
                final_outcome = rounds * hp_sum;
                break
            end
            rounds = rounds + 1;
        end
        if ~elf_died
            break
        end
    end
    fprintf('%d\n', final_outcome);
end

function [map, units, rows, cols, num_units] = parse_map(elf_power)
    fid = fopen('input.txt');
    lines = cellfun(@strtrim, textread('input.txt', '%s', 'whitespace', ''), 'UniformOutput', false);
    fclose(fid);
    rows = numel(lines);
    cols = max(cellfun(@length, lines));
    map = repmat(struct('kind', 1, 'x', 0, 'y', 0), rows, cols);
    units = repmat(struct('id', 0, 'kind', 1, 'hp', 200, 'power', 3, 'x', 0, 'y', 0, 'alive', true), 50, 1);
    num_units = 0;
    id = 0;
    for y = 1:rows
        line = lines{y};
        for x = 1:cols
            ch = line(x);
            switch ch
                case '#'
                    map(y, x).kind = 8;
                case '.'
                    map(y, x).kind = 1;
                case 'E'
                    map(y, x).kind = 1;
                    id = id + 1;
                    num_units = num_units + 1;
                    units(num_units) = struct('id', id, 'kind', 2, 'hp', 200, 'power', elf_power, 'x', x, 'y', y, 'alive', true);
                case 'G'
                    map(y, x).kind = 1;
                    id = id + 1;
                    num_units = num_units + 1;
                    units(num_units) = struct('id', id, 'kind', 4, 'hp', 200, 'power', 3, 'x', x, 'y', y, 'alive', true);
            end
            map(y, x).x = x;
            map(y, x).y = y;
        end
    end
    units = units(1:num_units);
end

function [combat_continues, hp_sum] = get_status(units, num_units)
    elves_exist = false;
    goblins_exist = false;
    hp_sum = 0;
    for i = 1:num_units
        if units(i).alive
            if units(i).kind == 2
                elves_exist = true;
            else
                goblins_exist = true;
            end
            hp_sum = hp_sum + units(i).hp;
        end
    end
    combat_continues = elves_exist && goblins_exist;
end

function [round_completed, elf_died, map, units, num_units] = tick(map, units, rows, cols, num_units, stop_on_elf_death)
    units = sort_units(units, num_units);
    elf_died = false;
    for i = 1:num_units
        if ~units(i).alive
            continue
        end
        current_unit = units(i);
        target_kind = 4;
        if current_unit.kind == 4
            target_kind = 2;
        end
        targets_exist = false;
        for j = 1:num_units
            if units(j).alive && units(j).kind == target_kind
                targets_exist = true;
                break
            end
        end
        if ~targets_exist
            round_completed = false;
            return
        end
        attack_target = [];
        min_hp = inf;
        offsets = [0 -1; -1 0; 1 0; 0 1];
        for k = 1:4
            nx = current_unit.x + offsets(k, 1);
            ny = current_unit.y + offsets(k, 2);
            if nx >= 1 && nx <= cols && ny >= 1 && ny <= rows
                for j = 1:num_units
                    if units(j).alive && units(j).kind == target_kind && units(j).x == nx && units(j).y == ny
                        if units(j).hp < min_hp
                            min_hp = units(j).hp;
                            attack_target = j;
                        end
                        break
                    end
                end
            end
        end
        if isempty(attack_target)
            [dist, path] = bfs(map, rows, cols, current_unit.x, current_unit.y);
            targets = [];
            target_count = 0;
            min_dist = inf;
            for j = 1:num_units
                if units(j).alive && units(j).kind == target_kind
                    for k = 1:4
                        nx = units(j).x + offsets(k, 1);
                        ny = units(j).y + offsets(k, 2);
                        if nx >= 1 && nx <= cols && ny >= 1 && ny <= rows && map(ny, nx).kind == 1 && dist(ny, nx) ~= -1
                            if dist(ny, nx) < min_dist
                                min_dist = dist(ny, nx);
                                target_count = 1;
                                targets = [nx ny];
                            elseif dist(ny, nx) == min_dist
                                target_count = target_count + 1;
                                targets = [targets; nx ny];
                            end
                        end
                    end
                end
            end
            if target_count > 0
                [~, idx] = sortrows(targets, [2 1]);
                chosen_target = targets(idx(1), :);
                current_step = chosen_target;
                while dist(current_step(2), current_step(1)) > 1
                    current_step = path(current_step(2), current_step(1), :);
                    if current_step(1) == -1
                        break
                    end
                end
                if current_step(1) ~= -1 && dist(current_step(2), current_step(1)) == 1
                    map(current_unit.y, current_unit.x).kind = 1;
                    units(i).x = current_step(1);
                    units(i).y = current_step(2);
                    map(units(i).y, units(i).x).kind = units(i).kind;
                    min_hp = inf;
                    attack_target = [];
                    for k = 1:4
                        nx = units(i).x + offsets(k, 1);
                        ny = units(i).y + offsets(k, 2);
                        if nx >= 1 && nx <= cols && ny >= 1 && ny <= rows
                            for j = 1:num_units
                                if units(j).alive && units(j).kind == target_kind && units(j).x == nx && units(j).y == ny
                                    if units(j).hp < min_hp
                                        min_hp = units(j).hp;
                                        attack_target = j;
                                    end
                                    break
                                end
                            end
                        end
                    end
                end
            end
        end
        if ~isempty(attack_target)
            units(attack_target).hp = units(attack_target).hp - units(i).power;
            if units(attack_target).hp <= 0
                units(attack_target).alive = false;
                map(units(attack_target).y, units(attack_target).x).kind = 1;
                if units(attack_target).kind == 2 && stop_on_elf_death
                    elf_died = true;
                    round_completed = true;
                    return
                end
            end
        end
    end
    live_units = 0;
    for k = 1:num_units
        if units(k).alive
            live_units = live_units + 1;
            units(live_units) = units(k);
        end
    end
    num_units = live_units;
    units = units(1:num_units);
    round_completed = true;
end

function [dist, path] = bfs(map, rows, cols, sx, sy)
    dist = -ones(rows, cols);
    path = repmat([-1 -1], rows, cols, 2);
    queue = [sx sy];
    dist(sy, sx) = 0;
    q_head = 1;
    offsets = [0 -1; -1 0; 1 0; 0 1];
    while q_head <= size(queue, 1)
        curr = queue(q_head, :);
        q_head = q_head + 1;
        for i = 1:4
            next_x = curr(1) + offsets(i, 1);
            next_y = curr(2) + offsets(i, 2);
            if next_x >= 1 && next_x <= cols && next_y >= 1 && next_y <= rows && map(next_y, next_x).kind == 1 && dist(next_y, next_x) == -1
                dist(next_y, next_x) = dist(curr(2), curr(1)) + 1;
                path(next_y, next_x, :) = [curr(1) curr(2)];
                queue = [queue; next_x next_y];
            end
        end
    end
end

function units = sort_units(units, num_units)
    if num_units <= 1
        return
    end
    [~, idx] = sortrows([units.y; units.x]');
    units = units(idx);
end

File: rust/day15_part2_2018.rs

use std::cmp::Ordering;
use std::collections::VecDeque;
use std::fs::File;
use std::io::{BufRead, BufReader};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Kind {
    Space,
    Elf,
    Goblin,
    Wall,
}

#[derive(Debug, Clone, Copy)]
struct Point {
    x: i32,
    y: i32,
}

#[derive(Debug, Clone)]
struct Unit {
    id: usize,
    kind: Kind,
    hp: i32,
    power: i32,
    x: i32,
    y: i32,
    alive: bool,
}

#[derive(Debug, Clone)]
struct State {
    map: Vec<Vec<Kind>>,
    units: Vec<Unit>,
    rows: i32,
    cols: i32,
}

const OFFSETS: [Point; 4] = [
    Point { x: 0, y: -1 },
    Point { x: -1, y: 0 },
    Point { x: 1, y: 0 },
    Point { x: 0, y: 1 },
];

fn parse_map(input: &str, elf_power: i32) -> State {
    let lines: Vec<_> = input.lines().collect();
    let rows = lines.len();
    let cols = lines[0].len();
    let mut map = vec![vec![Kind::Space; cols]; rows];
    let mut units = Vec::new();
    let mut unit_id = 0;

    for (y, line) in lines.iter().enumerate() {
        for (x, c) in line.chars().enumerate() {
            let kind = match c {
                '#' => Kind::Wall,
                '.' => Kind::Space,
                'E' => Kind::Elf,
                'G' => Kind::Goblin,
                _ => Kind::Space,
            };
            map[y][x] = kind;
            if kind == Kind::Elf || kind == Kind::Goblin {
                units.push(Unit {
                    id: unit_id,
                    kind,
                    hp: 200,
                    power: if kind == Kind::Elf { elf_power } else { 3 },
                    x: x as i32,
                    y: y as i32,
                    alive: true,
                });
                unit_id += 1;
            }
        }
    }

    State {
        map,
        units,
        rows: rows as i32,
        cols: cols as i32,
    }
}

fn get_status(units: &[Unit]) -> (bool, i32) {
    let mut elves = false;
    let mut goblins = false;
    let mut total_hp = 0;

    for unit in units {
        if unit.alive {
            total_hp += unit.hp;
            if unit.kind == Kind::Elf {
                elves = true;
            } else {
                goblins = true;
            }
        }
    }

    (elves && goblins, total_hp)
}

fn bfs(start: Point, map: &[Vec<Kind>], rows: i32, cols: i32) -> (Vec<Vec<i32>>, Vec<Vec<Point>>) {
    let mut dist = vec![vec![-1; cols as usize]; rows as usize];
    let mut path = vec![vec![Point { x: -1, y: -1 }; cols as usize]; rows as usize];
    let mut queue = VecDeque::new();

    dist[start.y as usize][start.x as usize] = 0;
    queue.push_back(start);

    while let Some(curr) = queue.pop_front() {
        for offset in &OFFSETS {
            let next = Point {
                x: curr.x + offset.x,
                y: curr.y + offset.y,
            };

            if next.x >= 0 && next.x < cols && next.y >= 0 && next.y < rows {
                let next_y = next.y as usize;
                let next_x = next.x as usize;
                if map[next_y][next_x] == Kind::Space && dist[next_y][next_x] == -1 {
                    dist[next_y][next_x] = dist[curr.y as usize][curr.x as usize] + 1;
                    path[next_y][next_x] = curr;
                    queue.push_back(next);
                }
            }
        }
    }

    (dist, path)
}

fn tick(state: &mut State, stop_on_elf_death: bool) -> (bool, bool) {
    state.units.sort_by(|a, b| {
        a.y.cmp(&b.y)
            .then_with(|| a.x.cmp(&b.x))
    });

    let mut elf_died = false;
    let mut i = 0;
    while i < state.units.len() {
        if !state.units[i].alive {
            i += 1;
            continue;
        }

        let current_unit = state.units[i].clone();
        let target_kind = if current_unit.kind == Kind::Elf {
            Kind::Goblin
        } else {
            Kind::Elf
        };

        let targets_exist = state
            .units
            .iter()
            .any(|u| u.alive && u.kind == target_kind);
        if !targets_exist {
            return (false, elf_died);
        }

        let mut attack_target: Option<usize> = None;
        let mut min_hp = i32::MAX;

        for offset in &OFFSETS {
            let nx = current_unit.x + offset.x;
            let ny = current_unit.y + offset.y;
            if nx >= 0 && nx < state.cols && ny >= 0 && ny < state.rows {
                for (j, unit) in state.units.iter().enumerate() {
                    if unit.alive
                        && unit.kind == target_kind
                        && unit.x == nx
                        && unit.y == ny
                        && unit.hp < min_hp
                    {
                        min_hp = unit.hp;
                        attack_target = Some(j);
                    }
                }
            }
        }

        if attack_target.is_none() {
            let (dist, path) = bfs(
                Point {
                    x: current_unit.x,
                    y: current_unit.y,
                },
                &state.map,
                state.rows,
                state.cols,
            );

            let mut targets = Vec::new();
            let mut min_dist = i32::MAX;

            for unit in &state.units {
                if unit.alive && unit.kind == target_kind {
                    for offset in &OFFSETS {
                        let nx = unit.x + offset.x;
                        let ny = unit.y + offset.y;
                        if nx >= 0 && nx < state.cols && ny >= 0 && ny < state.rows {
                            let ny_usize = ny as usize;
                            let nx_usize = nx as usize;
                            if state.map[ny_usize][nx_usize] == Kind::Space
                                && dist[ny_usize][nx_usize] != -1
                            {
                                if dist[ny_usize][nx_usize] < min_dist {
                                    min_dist = dist[ny_usize][nx_usize];
                                    targets.clear();
                                    targets.push(Point { x: nx, y: ny });
                                } else if dist[ny_usize][nx_usize] == min_dist {
                                    targets.push(Point { x: nx, y: ny });
                                }
                            }
                        }
                    }
                }
            }

            if !targets.is_empty() {
                targets.sort_by(|a, b| a.y.cmp(&b.y).then_with(|| a.x.cmp(&b.x)));
                let chosen_target = targets[0];
                let mut current_step = chosen_target;

                while dist[current_step.y as usize][current_step.x as usize] > 1 {
                    current_step = path[current_step.y as usize][current_step.x as usize];
                    if current_step.x == -1 {
                        break;
                    }
                }

                if current_step.x != -1 && dist[current_step.y as usize][current_step.x as usize] == 1
                {
                    state.map[current_unit.y as usize][current_unit.x as usize] = Kind::Space;
                    state.units[i].x = current_step.x;
                    state.units[i].y = current_step.y;
                    state.map[current_step.y as usize][current_step.x as usize] = current_unit.kind;

                    for offset in &OFFSETS {
                        let nx = state.units[i].x + offset.x;
                        let ny = state.units[i].y + offset.y;
                        if nx >= 0 && nx < state.cols && ny >= 0 && ny < state.rows {
                            for (j, unit) in state.units.iter().enumerate() {
                                if unit.alive
                                    && unit.kind == target_kind
                                    && unit.x == nx
                                    && unit.y == ny
                                    && unit.hp < min_hp
                                {
                                    min_hp = unit.hp;
                                    attack_target = Some(j);
                                }
                            }
                        }
                    }
                }
            }
        }

        if let Some(target_idx) = attack_target {
            state.units[target_idx].hp -= state.units[i].power;
            if state.units[target_idx].hp <= 0 {
                state.units[target_idx].alive = false;
                state.map[state.units[target_idx].y as usize][state.units[target_idx].x as usize] =
                    Kind::Space;
                if state.units[target_idx].kind == Kind::Elf && stop_on_elf_death {
                    elf_died = true;
                    return (true, elf_died);
                }
            }
        }

        i += 1;
    }

    state.units.retain(|u| u.alive);
    (true, elf_died)
}

fn main() {
    let file = File::open("input.txt").unwrap();
    let reader = BufReader::new(file);
    let input = reader
        .lines()
        .map(|l| l.unwrap())
        .collect::<Vec<_>>()
        .join("\n");

    let mut elf_power = 3;
    let mut final_outcome = 0;

    loop {
        elf_power += 1;
        let mut state = parse_map(&input, elf_power);
        let mut rounds = 0;
        let mut elf_died = false;

        loop {
            let (combat_continues, hp_sum) = get_status(&state.units);
            if !combat_continues {
                final_outcome = rounds * hp_sum;
                break;
            }

            let (round_completed, died) = tick(&mut state, true);
            elf_died = died;

            if elf_died {
                break;
            }

            if !round_completed {
                let (_, hp_sum) = get_status(&state.units);
                final_outcome = rounds * hp_sum;
                break;
            }

            rounds += 1;
        }

        if !elf_died {
            break;
        }
    }

    println!("{}", final_outcome);
}
